### PR TITLE
Adding table get_file to retrieve contents of a file

### PR DESCRIPTION
**Pull request by javuto**
This table retrieves the content of a regular file and copies it to a folder, specified by the new flag, getfile_path. The file will have a non predictable name.

### CODE REVIEW

**Comment by osqueryer** [comment]
Build origin/pr/1917/merge (Job: 2461) failed :-1:

File: specs/utility/get_file.table [hunk 1]
```diff
@@ -0,0 +1,18 @@
+table_name("get_file")
+description("File contents extraction table.")
+schema([
+    Column("path", TEXT, "Absolute file path", required=True),
+    Column("filename", TEXT, "Name portion of file path"),
+    Column("directory", TEXT, "Directory of file(s)", required=True),
+    Column("size", BIGINT, "Size of file in bytes"),
+    Column("offset", BIGINT, "Size of file in bytes"),
```
Comment (friedbutter) [inline specs/utility/get_file.table]: 'offset' description is copied from 'size' :p

File: osquery/core/init.cpp [hunk 2]
```diff
@@ -168,6 +168,7 @@ DECLARE_bool(config_check);
 DECLARE_bool(config_dump);
 DECLARE_bool(database_dump);
 DECLARE_string(database_path);
+DECLARE_string(getfile_path);
```
Comment (theopolis) [inline osquery/core/init.cpp]: I'm hesitant about using an explicit path here. Perhaps making this default opaque to people and just use a secured temp folder is better.

File: osquery/core/init.cpp [hunk 3]
```diff
@@ -261,6 +262,16 @@ Initializer::Initializer(int& argc, char**& argv, ToolType tool)
       if (Flag::isDefault("database_path")) {
         osquery::FLAGS_database_path = homedir + "/shell.db";
       }
+      if (Flag::isDefault("getfile_path")) {
+        auto getfiledir = homedir + "/getfiles";
```
Comment (theopolis) [inline osquery/core/init.cpp]: Let's be safe, but not platform independent and use: [`mkdtemp`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/mkdtemp.html).

File: osquery/core/init.cpp [hunk 4]
```diff
@@ -261,6 +262,16 @@ Initializer::Initializer(int& argc, char**& argv, ToolType tool)
...
+        if (osquery::pathExists(getfiledir).ok() ||
+            boost::filesystem::create_directory(getfiledir, ec)) {
+          osquery::FLAGS_getfile_path = getfiledir;
+        } else {
+          LOG(INFO) << "Cannot access or create osquery getfile directory";
+          FLAGS_getfile_path = "/tmp";
```
Comment (theopolis) [inline osquery/core/init.cpp]: image

File: osquery/core/init.cpp [hunk 5]
```diff
@@ -268,6 +279,7 @@ Initializer::Initializer(int& argc, char**& argv, ToolType tool)
       LOG(INFO) << "Cannot access or create osquery home directory";
       FLAGS_disable_extensions = true;
       FLAGS_database_path = "/dev/null";
+      FLAGS_getfile_path = "/tmp";
```
Comment (theopolis) [inline osquery/core/init.cpp]: See above

File: osquery/tables/utility/get_file.cpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+const std::map<fs::file_type, std::string> kTypeNames{
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: Maybe put this into `./include/filesystem.h`:
```cpp
extern const std::map<fs::file_type, std::string> kTypeNames;
```

 and the definition in `./osquery/filesystem/filesystem.cpp`, don't duplicate: https://github.com/facebook/osquery/blob/master/osquery/tables/utility/file.cpp#L24

File: osquery/tables/utility/get_file.cpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    {fs::block_file, "block"},
+    {fs::character_file, "character"},
+    {fs::fifo_file, "fifo"},
+    {fs::socket_file, "socket"},
+    {fs::type_unknown, "unknown"},
+    {fs::status_error, "error"},
+};
+
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__linux__)
+const char token_charset[] =
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: Name these appropriately (with a camel-cast `kTokenCharset`), or perhaps move them into the `genFileToken` method and define them `static`.

File: osquery/tables/utility/get_file.cpp [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  return token;
+}
+
+void genGetFileContent(const fs::path& path,
+                 const fs::path& parent,
+                 const std::string& pattern,
+                 QueryData& results) {
+  // Must provide the path, filename, directory separate from boost path->string
+  // helpers to match any explicit (query-parsed) predicate constraints.
+  struct stat file_stat, link_stat;
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: A lot of this code is repeated from `readFile` and from `./osquery/tables/utility/file.cpp`, you can use other tables inline:
```cpp
auto results = SQL::selectAllFrom("file", "path", EQUALS, path.string());
```

File: osquery/tables/utility/get_file.cpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    r["type"] = "unknown";
+  }
+
+  r["path"] = path.string();
+  r["filename"] = path.filename().string();
+  r["directory"] = parent.string();
+  r["size"] = BIGINT(file_stat.st_size);
+  r["offset"] = BIGINT(0);
+
+  std::string getfile_path = Flag::getValue("getfile_path").c_str();
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: Why construct a `std::string` from the `const char*` accessor?

File: osquery/tables/utility/get_file.cpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  auto token = genFileToken(12);
+  auto token_path = getfile_path + "/" + token;
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: Might be overkill but:
```cpp
auto token_path = fs::path(getfile_path) / token;
```

Then when you need the string flattened: `token_path.string()`.

File: osquery/tables/utility/get_file.cpp [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+  r["token"] = TEXT(token_path);
+
+  fs::copy_file(path.string().c_str(), token_path, ec);
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: You have a `fs::path` object: `path`, just pass it to `fs::copy_file`.

File: osquery/tables/utility/get_file.cpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+      }
+    } catch (const fs::filesystem_error& e) {
+      continue;
+    }
+  }
+
+  return results;
+}
+}
+}
```
Comment (theopolis) [inline osquery/tables/utility/get_file.cpp]: Add newline

File: tools/deployment/osquery.example.conf [hunk 13]
```diff
@@ -36,6 +36,9 @@
     // and query results differentials. See also 'use_in_memory_database'.
     //"database_path": "/var/osquery/osquery.db",
 
+    // A filesystem path for storing the files retrieved with table get_file.
+    //"getfile_path": "/var/osquery/getfiles",
```
Comment (theopolis) [inline tools/deployment/osquery.example.conf]: Na, no need to add the reference here. Add it to https://github.com/facebook/osquery/blob/master/docs/wiki/installation/cli-flags.md

File: osquery/core/init.cpp (hunk 2 above)
Comment (javuto) [inline osquery/core/init.cpp]: I agree we can add this later if there is a need for it

**Comment by javuto** [comment]
This was open way too long, I will open a new pull request to handle the comments and implement this functionality

**Pull request closed without merging**